lib/service/version.go

```diff
@@ -0,0 +1,95 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package service
+
+import (
+	"errors"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	// majorVersionConstraint is the major version constraint when previous major version must be
+	// present in the storage, if not - we must refuse to start.
+	majorVersionConstraint = 18
+)
+
+var (
+	// errMajorVersionUpgrade default error for restricting upgrade major version.
+	errMajorVersionUpgrade = errors.New(`Teleport supports upgrade only one major version behind. ` +
+		`See: https://goteleport.com/docs/upgrading/overview/#component-compatibility`)
+)
+
+// validateAndUpdateTeleportVersion validates that the major version persistent in the backend
+// meets our upgrade compatibility guide.
+func (process *TeleportProcess) validateAndUpdateTeleportVersion(currentVersion string) error {
+	currentMajor, err := getMajorVersion(currentVersion)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	st, err := process.storage.GetState(process.GracefulExitContext(), types.RoleInstance)
+	if trace.IsNotFound(err) {
+		// If the Instance state has not been found, it means this is the first launch,
+		// and we don't need to check the version since this is the initial launch.
+		return nil
```

Comment:
I think we need to test a scenario when local process state gets removed, for whatever reason (shouldn't normally happen but people can get their clusters in all sorts of weird states). What would happen in this case? This will not be considered "first time start" because that is determined based on the presence of CA IIRC, will we re-init the process state with correct version? I'm mostly worried about scenario where it would get into some state which we'll have to recover manually which we need to avoid.

Reply:
For instance if we remove process database completely or specifically this `/states/instance/state` item, first we should go in this version check so `/states/instance/state` not created yet, this item managed by `process.connect` and going to be created after validation. Currently looking to in integration tests, to cover this case

Worst case scenario might be, if the user manually change `local_version` of `/states/instance/state` item in database

Reply:
Was testing couple of scenarios with integration tests, found that in configuration when we enable only auth service
```golang
	cfg.Auth.Enabled = true
	cfg.Proxy.Enabled = false
	cfg.SSH.Enabled = false
```
We never register the instance state in the process storage
https://github.com/gravitational/teleport/blob/dc1199636558f85ed275512175a12207a8034d44/lib/service/service.go#L2564-L2574
this leads to skipping version check in auth service

Reply:
@fspmarshall Any advice here?